clippy_lints/src/uninit_vec.rs

```diff
@@ -0,0 +1,174 @@
+use clippy_utils::diagnostics::span_lint_and_note;
+use clippy_utils::{is_uninit_ty_valid, match_def_path, path_to_local_id, paths, peel_hir_expr_while, SpanlessEq};
+use rustc_hir::def::Res;
+use rustc_hir::{Block, Expr, ExprKind, HirId, PatKind, Stmt, StmtKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::Span;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for the creation of uninitialized `Vec<T>` by calling `set_len()`
+    /// immediately after `with_capacity()` or `reserve()`.
+    ///
+    /// ### Why is this bad?
+    /// It creates `Vec<T>` that contains uninitialized data, which leads to an
+    /// undefined behavior with most safe operations.
+    /// Notably, using uninitialized `Vec<u8>` with generic `Read` is unsound.
+    ///
+    /// ### Example
+    /// ```rust,ignore
+    /// let mut vec: Vec<u8> = Vec::with_capacity(1000);
+    /// unsafe { vec.set_len(1000); }
+    /// reader.read(&mut vec); // undefined behavior!
+    /// ```
+    /// Use an initialized buffer:
+    /// ```rust,ignore
+    /// let mut vec: Vec<u8> = vec![0; 1000];
+    /// reader.read(&mut vec);
+    /// ```
+    /// Or, wrap the content in `MaybeUninit`:
+    /// ```rust,ignore
+    /// let mut vec: Vec<MaybeUninit<T>> = Vec::with_capacity(1000);
+    /// unsafe { vec.set_len(1000); }
+    /// ```
+    pub UNINIT_VEC,
+    correctness,
+    "Vec with uninitialized data"
+}
+
+declare_lint_pass!(UninitVec => [UNINIT_VEC]);
+
+impl<'tcx> LateLintPass<'tcx> for UninitVec {
+    fn check_block(&mut self, cx: &LateContext<'tcx>, block: &'tcx Block<'_>) {
+        for w in block.stmts.windows(2) {
```

Comment:
> This only catches cases where `Vec::with_capacity(..);` is directly followed by `vec.set_len(..);`

This was an intentional decision to reduce false positives. The statements between them might be a legit initialization:
```rust
self.vec.reserve(1000);
self.init_vec(); // correctly initializes `self.vec`, e.g., by using `mem::write()`
self.vec.set_len(1000); // this call is now correct
```

For statistics, checking directly adjacent statements are enough to detect most of incorrect usage of `set_len()`. We have a list of 30 soundness bugs that fall into this category, and 28 of them can be caught with the current implementation. The other two patterns that are not detected are:
```rust
let mut a = Vec::with_capacity(len);
let mut b = Vec::with_capacity(len);
unsafe {
    a.set_len(len);
    b.set_len(len);
}
```
and
```rust
let mut vec = Vec::with_capacity(len);
unsafe {
    vec.set_len(len);
    another_unsafe_fn(&mut vec);
}
```

I think I can slightly modify the block handling logic to detect the second case. Then, it would detect 29 out of 30 cases. Would that be good enough, or would you still prefer me to try detecting the first case?

My concern is that implementation complexity. (1) The block handling logic would become complicated because `with_capacity()/reserve()` and `set_len()` might not be in the same block due to `unsafe` blocks. (2) The visitor needs to detect whether the target `Vec` is saved to a local variable or to the field of a struct. Only the first case would be safe to detect with the visitor due to the false positives mentioned above.

Reply:
Okay, from the example, I see what you mean. Cases where the `Vec` is stored inside a struct or tuple would not be caught, that's correct. Most lints related to sequential usage only work for locals, as far as I know. Do you know how many of the mentioned cases, used the vector as a part of a struct or tuple?

> (1) The block handling logic would become complicated because `with_capacity()/reserve()` and `set_len()` might not be in the same block due to `unsafe` blocks.

The Visitor or LintPass both also visit nested blocks. They would therefore also work with unsafe blocks or more complicated logic.

> (2) The visitor needs to detect whether the target `Vec` is saved to a local variable or to the field of a struct. Only the first case would be safe to detect with the visitor due to the false positives mentioned above.

Hmm, I think the best way to handle this would be to track the local that the vector is stored in and check for the first usage. This would mean the vector itself if it's a local value or the struct that the vector is stored in. This shouldn't add too much additional complexity (I believe...).

However, I'm not totally sure on this. @flip1995 could you maybe take a look at this? :upside_down_face:

Reply:
> Do you know how many of the mentioned cases, used the vector as a part of a struct or tuple?

They were roughly 15% of the bugs.

> The Visitor or LintPass both also visit nested blocks. They would therefore also work with unsafe blocks or more complicated logic.

> Hmm, I think the best way to handle this would be to track the local that the vector is stored in and check for the first usage.

Oh, now I see what you were suggesting by "The second option is to store the ID of the value in `Self`". I was initially confused by `vec_init_then_push` example because it handles neither nested blocks nor in-between statements. Is it correct that you were providing it as an example of how to use a lint pass as a visitor while maintaining the state in `self`, not necessarily as an example for implementing the visitor logic itself?

I understand what I need to do in that case. However, I think I might still prefer the current version over the version that checks in-between statements due to the simplicity and possible false positives. Would you mind if I open a separate PR for the visitor version that checks in-between statements? Then maybe we can compare the two versions and merge the one that looks better. I'll start working on it after addressing the comments about the lint messages below.

In addition, could you recommend a lint example that handles in-between statements? I was thinking of using `HashMap` that maps the `HirId` of `Local` to its tracked status, and I was wondering if there is something that I can reuse from the existing lints. Thanks!

Reply:
If we use a separate visitor struct, we can create a visitor for each local variable (e.g., the use of `id` field in [needless_collect](https://github.com/rust-lang/rust-clippy/blob/a4a3f61c9fed1e7a319b97637bcf72c6a572176c/clippy_lints/src/loops/needless_collect.rs#L89-L93) lint) so `HashMap` wouldn't be needed in this case. However, it would have a quadratic performance in terms of # locals/block size. This seems fine since a visitor will be only created for each `with_capacity()/reserve()`.

If we use `LintPass` itself as a visitor and handle all variables in a single loop, it would need to manage a state for each variable and might need a data structure like `HashMap`.

Reply:
That makes sense. It would be interesting to do some profiling to compare a `Visitor` implementation to a `LintPass` implementation. This lint can choose either way, due to the fact that the actual logic for each expression is relatively simple. :upside_down_face: